hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala

```diff
@@ -1,25 +1,88 @@
 package is.hail.backend.service
 
-import java.io.{PrintWriter, StringWriter}
+import java.io.{DataOutputStream, FileInputStream, FileOutputStream, ObjectInputStream, ObjectOutputStream, PrintWriter, RandomAccessFile, StringWriter}
 
 import is.hail.annotations.{Region, UnsafeRow}
 import is.hail.asm4s._
-import is.hail.backend.{Backend, BroadcastValue}
+import is.hail.backend.{Backend, BackendContext, BroadcastValue}
 import is.hail.expr.JSONAnnotationImpex
 import is.hail.expr.ir.lowering.{DArrayLowering, LowerDistributedSort, LowererUnsupportedOperation, LoweringPipeline, TableStage}
 import is.hail.expr.ir.{Compile, ExecuteContext, IR, IRParser, MakeTuple, SortField}
 import is.hail.types.physical.{PBaseStruct, PType}
-import is.hail.io.fs.GoogleStorageFS
+import is.hail.io.fs.{FS, GoogleStorageFS}
+import is.hail.services.batch_client.BatchClient
 import is.hail.utils._
+import org.apache.commons.io.IOUtils
+import org.apache.log4j.LogManager
 import org.json4s.JsonAST.{JArray, JBool, JInt, JObject, JString}
 import org.json4s.jackson.JsonMethods
 
 import scala.collection.mutable
 import scala.reflect.ClassTag
 
+object ServiceBackendUtils {
+  def getQueryFS(): FS = {
+    var keyFile = System.getenv("HAIL_QUERY_GSAKEY_FILE")
+    if (keyFile == null)
+      keyFile = "/gsa-key/key.json"
+
+    using(new FileInputStream(keyFile)) { is =>
+      new GoogleStorageFS(IOUtils.toString(is))
+    }
+  }
+}
+
+object Worker {
+  def main(args: Array[String]): Unit = {
+    if (args.length != 2)
+      throw new IllegalArgumentException(s"expected one argument, not: ${ args.length }")
+
+    val root = args(0)
+    val i = args(1).toInt
+
+    val fs = ServiceBackendUtils.getQueryFS()
+
+    val f = using(new ObjectInputStream(fs.openNoCompression(s"$root/f"))) { is =>
+      is.readObject().asInstanceOf[(Array[Byte], Int) => Array[Byte]]
+    }
+
+    var offset = 0L
+    var length = 0
+
+    using(fs.openNoCompression(s"$root/context.offsets")) { is =>
+      is.seek(i * 12)
+      offset = is.readLong()
+      length = is.readInt()
+    }
+
+    println(s"offset $offset length $length")
+
+    val context = using(fs.openNoCompression(s"$root/contexts")) { is =>
+      is.seek(offset)
+      val context = new Array[Byte](length)
+      is.readFully(context)
+      context
+    }
+
+    val result = f(context, i)
+
+    using(fs.createNoCompression(s"$root/result.$i")) { os =>
+      os.write(result)
+    }
+  }
+}
+
+class ServiceBackendContext(
+  val username: String,
+  val sessionID: String,
+  val billingProject: String
+) extends BackendContext with Serializable
```

Comment:
Should this be serializable? It contains the most secret thing a user has: the sessionID. I almost want to override toString to specifically not print the sessionID just to head off someone unintentionally doing that in the future.

Reply:
Yeah, you're right.  I'll try without it and see if/where it fails.  I think you're right tho, it shouldn't have to leave the driver.

It isn't a case class so toString won't print the contents:

```
scala> class A(val sensitive: Int)
defined class A

scala> val a = new A(5)
a: A = A@5fb97279
```